### PR TITLE
[SPARK-17721][MLlib][ML] Fix for multiplying transposed SparseMatrix with SparseVector

### PR DESCRIPTION
**Pull request by bwahlgreen**
## What changes were proposed in this pull request?
- changes the implementation of gemv with transposed SparseMatrix and SparseVector both in mllib-local and mllib (identical)
- adds a test that was failing before this change, but succeeds with these changes.

The problem in the previous implementation was that it only increments `i`, that is enumerating the columns of a row in the SparseMatrix, when the row-index of the vector matches the column-index of the SparseMatrix. In cases where a particular row of the SparseMatrix has non-zero values at column-indices lower than corresponding non-zero row-indices of the SparseVector, the non-zero values of the SparseVector are enumerated without ever matching the column-index at index `i` and the remaining column-indices i+1,...,indEnd-1 are never attempted. The test cases in this PR illustrate this issue.
## How was this patch tested?

I have run the specific `gemv` tests in both mllib-local and mllib. I am currently still running `./dev/run-tests`. 
## ___

As per instructions, I hereby state that this is my original work and that I license the work to the project (Apache Spark) under the project's open source license.

Mentioning @dbtsai, @viirya and @brkyvz whom I can see have worked/authored on these parts before.

### CODE REVIEW

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala [hunk 1]
```diff
@@ -638,12 +638,18 @@ private[spark] object BLAS extends Serializable {
...
         var sum = 0.0
         var k = 0
-        while (k < xNnz && i < indEnd) {
+        while (i < indEnd && k < xNnz) {
           if (xIndices(k) == Acols(i)) {
             sum += Avals(i) * xValues(k)
+            k += 1
+            i += 1
+          }
+          else if (xIndices(k) < Acols(i)) {
```
Comment (srowen) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala]: Nit: pull the elses up onto the previous line. Does this affect the non-transposed case below too? I believe your change is indeed correct and the tests demonstrate that.

Comment (bwahlgreen) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala]: elses on previous line, done. I must admit I haven't gone through the non-transpose case in as much detail, but that does not seem affected by this bug. Should I perhaps add a test that shows the non-transpose case is/remains unaffected?

Comment (srowen) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala]: I think that would be ideal if it's not already covered, just to be sure. It ought to be a straightforward modification of a bit of test code you just added.

Comment (bwahlgreen) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala]: check

**Comment by jkbradley** [comment]
ok to test

**Comment by jkbradley** [comment]
LGTM
I'll merge this with master and plan to backport it to branches for 2.0, 1.6, and 1.5.
Thanks!

**Comment by jkbradley** [comment]
@bwahlgreen  Actually since mllib-local was not in 1.6 or 1.5, would you mind sending a backport PR against branch-1.6?  Thank you!

File: mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala [hunk 2]
```diff
@@ -638,12 +638,16 @@ private[spark] object BLAS extends Serializable {
         val indEnd = Arows(rowCounter + 1)
         var sum = 0.0
         var k = 0
-        while (k < xNnz && i < indEnd) {
+        while (i < indEnd && k < xNnz) {
```
Comment (viirya) [inline mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala]: nit: an unnecessary change?

**Comment by viirya** [comment]
LGTM although it was merged.
